src/sciline/pipeline.py

```diff
@@ -206,25 +395,111 @@ def build(self, tp: Type[T], /) -> Graph:
         ----------
         tp:
             Type to build the graph for.
+        search_param_tables:
+            Whether to search parameter tables for concrete keys.
         """
         graph: Graph = {}
-        stack: List[Type[T]] = [tp]
+        stack: List[Union[Type[T], Item[T]]] = [tp]
         while stack:
             tp = stack.pop()
+            if search_param_tables and tp in self._param_series:
+                graph[tp] = (_param_sentinel, (self._param_series[tp],))
+                continue
+            if get_origin(tp) == Series:
+                graph.update(self._build_series(tp))  # type: ignore[arg-type]
+                continue
             provider: Callable[..., T]
             provider, bound = self._get_provider(tp)
             tps = get_type_hints(provider)
-            args = {
-                name: _bind_free_typevars(t, bound=bound)
+            args = tuple(
+                _bind_free_typevars(t, bound=bound)
                 for name, t in tps.items()
                 if name != 'return'
-            }
+            )
             graph[tp] = (provider, args)
-            for arg in args.values():
+            for arg in args:
                 if arg not in graph:
                     stack.append(arg)
         return graph
 
+    def _build_series(self, tp: Type[Series[KeyType, ValueType]]) -> Graph:
```

Comment:
Quite a long and deeply nested function. Can you split it up, e.g., along the steps?

Also, I don't understand it. We may need to go through it together. But refactoring may already help.

Reply:
Yes, this is the part that I wanted to do the walk-through for.

Reply:
I completely rewrote this, moving the functionality mostly into the "grouper" classes. I think it is more readable now, please have a look @jl-wynen .